salt/cache/etcd_cache.py

```diff
@@ -182,7 +195,10 @@ def _walk(r):
     r: etcd.EtcdResult
     """
     if not r.dir:
-        return [r.key.split("/", 3)[3]]
+        if r.key.endswith(_tstamp_suffix):
+            return []
+        else:
+            return [r.key.rsplit("/", 1)[-1]]
```

Comment:
Previously ~on X-Files~ in this module we required a specific nesting - in particular IIRC it wanted `1/2/3` for example. This is now fixed to where arbitrary nesting will work correctly in etcd. Including something absolutely ridiculous (see the tests for an example).

Reply:
This looks like we would be breaking backwards capabilities. Would this break users existing setups if using the etcd cache?

Reply:
I'm not sure I follow - before it did /not/ allow arbitrary nesting (and localfs used os.walk, IIRC)

It only functioned correctly with precisely 3 levels of nesting - anything else would produce incorrect results at other levels of nesting.

I can double check, but... It was definitely inconsistent behavior, and only worked by accident, basically.

Reply:
First off: Thank you for the challenge! It's actually be super helpful for me to go back and double check this :+1: 

I was pretty confident that this is doing the right thing but now I'm absolutely positive.

Consider the default salt setup and run the following:

```
salt-run cache.store yourminion thing "some data goes here"
salt-run cache.list yourminion
# output:
# - thing
salt-run cache.fetch yourminion thing
# output:
# some data goes here
```

Now change the master config:

```
cache: etcd
etcd.host: 172.17.0.2   # my docker container
etcd.port: 2379
etcd.protocol: http
etcd.path_prefix: /foo/bar/bang
```

re-run the same commands and the list one will produce the output:

```
bang/yourminion/thing
```

Which is uh... not the same as localfs. From that point alone, I'd say that the existing behavior is buggy.

As far as tripping up anyone who's currently using etcd caching, I don't *think* that there will be any issues, because will store and retrieve the values the same, but that's basically because from etcd's perspective `'/this' + '/that/path'` is no different from `'/this/that' + '/path'`. In other words I'm pretty sure it's just by accident that it works at all. 

So we're definitely correcting a bug with the way that list works with the runner. I'm guessing the only reason we haven't seen any problems is because nobody ever hit this. Though [this issue might have something to do with it](https://github.com/saltstack/salt/issues/56175).

After all this poking around, I'm also fairly confident that we're not going to blow anything up - I spun up this container running the master branch version of salt and got some data cached, then I changed to this version and from what I could see by running curl commands and trying to fetch cache data there weren't any changes between what was stored. It's *possible* that I missed something, and I wouldn't be against putting another note of caution in the changelog if we think it'd be helpful.